examples/finetune_resnet/components.py

```diff
@@ -0,0 +1,138 @@
+import imp
+import json
+from typing import Any, Dict, List, Optional
+
+import datasets
+import torch
+from cached_path import cached_path
+from PIL import Image
+from torch import nn
+from torch.optim import Adam
+from torchvision import models, transforms
+
+from tango import Format, JsonFormat, Step
+from tango.integrations.torch import DataCollator, Model, Optimizer
+
+# Register the Adam optimizer as an `Optimizer` so we can use it in the train step.
+Optimizer.register("torch_adam")(Adam)
+
+
+# Wrapper class around the pre-trained ResNet-18 model that modifies the final layer.
+@Model.register("resnet_ft")
+class ResNetWrapper(Model):
+    def __init__(self, num_classes: int, feature_extract: bool, use_pretrained: bool):
+        super().__init__()
+        self.model_ft = models.resnet18(pretrained=use_pretrained)
+        self.set_parameter_requires_grad(self.model_ft, feature_extract)
+        num_features = self.model_ft.fc.in_features
+        self.model_ft.fc = nn.Linear(num_features, num_classes)
+        self.loss_fn = nn.CrossEntropyLoss()
+
+    def set_parameter_requires_grad(self, model: models, feature_extracting: bool):
+        if feature_extracting:
+            for param in model.parameters():
+                param.requires_grad = False
+
+    def forward(
+        self, image: torch.Tensor, label: Optional[torch.Tensor] = None
+    ) -> Dict[str, torch.Tensor]:
+        output = self.model_ft(image)
+        if label is None:
+            return {"pred": torch.argmax(output)}
+        loss = self.loss_fn(output, label)
```

Comment:
We should also calculate the accuracy - at least when not in "training" model (`self.training` is `False`) - and return that along with the loss.